hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/upgrade/ReconLayoutFeature.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hadoop.ozone.recon.upgrade;
+
+/**
+ * Enum representing Recon layout features with their version, description,
+ * and associated upgrade action to be executed during an upgrade.
+ */
+public enum ReconLayoutFeature {
+  INITIAL_VERSION(0, "Initial Layout Version", new InitialLayoutUpgradeAction()),
```

Comment:
We do not need binding the Action for every version, when required, can be dinamically added on scan, like done for OM - org.apache.hadoop.ozone.om.upgrade.QuotaRepairUpgradeAction

Reply:
Thanks for the comment @sumitagrawl 

In ReconLayoutFeature, upgrade actions are embedded directly within the enum values, making the process simple and straightforward. Each feature (e.g., FEATURE_1) has its corresponding action (e.g., Feature1UpgradeAction). During an upgrade, the system compares the current MLV (Metadata Layout Version) and SLV (Software Layout Version). If the MLV is outdated, the associated action is executed, and the MLV is updated. Since Recon does not have many upgrade phases, I decided to omit additional phases, keeping the upgrade process simple and easy to manage.

On the other hand, OMLayoutFeature introduces a more complex process. Actions are not embedded but dynamically registered and associated with different phases, like FIRST_RUN_ON_UPGRADE and ON_FINALIZE. The system scans for upgrade actions and links them to features based on annotations as different actions can be triggered at different stages of the upgrade, making it suitable for more complex upgrade scenarios.

In summary, Recon uses a simpler, more direct approach by omitting multiple upgrade phases, while OM provides flexibility through phased upgrades, allowing it to handle more intricate upgrade processes.

Should I shift the approach then?

Reply:
There is a plan to support prepare and finalize -- may not be part of this PR.
if no action, either you can define it as functional interface with empty lamda implementation for Version "0".

private static final ReconUpgradeAction NO_FINALIZE_ACTION = () -> {};
INITIAL_VERSION(0, "Initial Layout Version", NO_FINALIZE_ACTION),

This can be refactored later on based on required support.